rpmio/rpmpgp.c

```diff
@@ -1057,38 +1055,135 @@ unsigned int pgpDigParamsAlgo(pgpDigParams digp, unsigned int algotype)
     return algo;
 }
 
+static pgpDigParams pgpDigParamsNew(uint8_t tag)
+{
+    pgpDigParams digp = xcalloc(1, sizeof(*digp));
+    digp->tag = tag;
+    return digp;
+}
+
+static void hashKey(DIGEST_CTX hash, const struct pgpPkt *pkt)
+{
+    uint8_t head[] = {
+	0x99,
+	(pkt->blen >> 8),
+	(pkt->blen     ),
+    };
+
+    rpmDigestUpdate(hash, head, 3);
+    rpmDigestUpdate(hash, pkt->body, pkt->blen);
+}
+
+static void hashUser(DIGEST_CTX hash, const struct pgpPkt *pkt)
+{
+    uint8_t head[] = {
+	0xb4,
+	(pkt->blen >> 24),
+	(pkt->blen >> 16),
+	(pkt->blen >>  8),
+	(pkt->blen      ),
+    };
+    rpmDigestUpdate(hash, head, 5);
+    rpmDigestUpdate(hash, pkt->body, pkt->blen);
+}
+
+static int pgpVerifySelf(pgpDigParams key, pgpDigParams selfsig,
+			const struct pgpPkt *all, int i)
+{
+    int xx = -1;
+    DIGEST_CTX hash = rpmDigestInit(selfsig->hash_algo, 0);
+    hashKey(hash, &all[0]);
+
```

Comment:
You might want to verify the package types of the package you're hashing, i.e. `all[i-1]` should be of the required type.

Reply:
Does it actually matter though? I mean, if the type is wrong then the hash will also be wrong and simply fail to pass? Okay, as a belt and suspenders kind of thing I suppose...

Reply:
I also think that it does not matter, but it's the right thing to do for security relevant code ;-)

Anyway, for the _CERT verification you can't use all[i-1] as there may be multiple signatures.

Reply:
Changed it to check for the types and more errors overall, userid lookup in the CERT case still missing.